dhall/src/Dhall.hs

```diff
@@ -900,6 +909,103 @@ instance (Interpret a, Interpret b) => Interpret (a, b)
 auto :: Interpret a => Type a
 auto = autoWith defaultInterpretOptions
 
+newtype DhallFix f t = DhallFix { unDhallFix :: t }
+newtype FixVar = FixVar { unFixVar :: (Expr Src X) }
+
+-- This is a helper instance used from implementation of Interpret for DhallFix
+instance Interpret FixVar where
+    autoWith _ = Type {..}
+        where
+        extract (App _ e0) = pure (FixVar e0)
+        extract e          = typeError expected e
+
+        expected = "t"
+
+type InterpretFix f t =
+   ( Traversable f
+   , Corecursive t
+   , Interpret (f FixVar)
+   , Base t ~ f
+   )
+
+instance InterpretFix f t => Interpret (DhallFix f t) where
+    autoWith opts = Type { extract = extr, expected = expe }
+        where
+        extr (Lam _ _ (Lam _ _ e)) = DhallFix <$> buildF e
+        extr e                     = typeError expe e
+
+        -- buildF = elgot f g <=< extract (auto @FixVar)
+        viaMonadic a b x = fromMonadic $ toMonadic . a =<< toMonadic (b x)
+        buildF = elgot f g `viaMonadic` extract (auto :: Type FixVar)
+
+        g = repack . extract xF_t . unFixVar
+        repack (Failure e) = Left (Failure e)
+        repack (Success s) = Right (Free s)
+        f (Free s) = Data.Functor.Foldable.embed <$> sequenceA s
+        f (Pure x) = typeError expe x
+
+        expe = Pi "t" (Dhall.Core.Const Dhall.Core.Type) -- = ∀(t : Type)
+             $ Pi "_" (Pi "_" (expected xF_t) "t")       -- → ∀(F t → t)
+             $ "t"                                       -- → t
+
+        xF_t = autoWith opts :: Type (f FixVar)
+
+{-| `autoWithFix` allows you to implement `Interpret` instance for a recursive
+    data type. For example, given a Haskell expression:
+
+  > data Expr
+  >    = Lit Natural
+  >    | Add Expr Expr
+  >    | Mul Expr Expr
+  >    deriving (Generic, Show)
+  >
+  > makeBaseFunctor ''Expr
+  > deriving instance Generic (ExprF a)
+  > instance Interpret a => Interpret (ExprF a)
+  > instance Interpret Expr where autoWith = autoWithFix
+
+    And a dhall file:
+
+  > -- expr.dhall
+  >
+  > -- F-Algebra declaration.
+  > let ExprF
+  >   = λ(t : Type) →
+  >   < LitF : { _1 : Natural }
+  >   | AddF : { _1 : t, _2 : t }
+  >   | MulF : { _1 : t, _2 : t }
+  >   >
+  >
+  > -- Helper function.
+  > let Expr
+  >   = λ(t : Type)
+  >   → λ(fix : ExprF t → t)
```

Comment:
Minor suggestion: Perhaps name this bound variable in the example `recurse` instead of `fix`

Reply:
So my counter here is: 1. fix is shorter. 2. Since this documentation is targeted at Haskellers it will give them additional intuition by analogy to `Fix`, e.g.:

```haskell
Fix (AddF (Fix (LitF 8)) (Fix (LitF 2)))
```

Reply:
Then perhaps use capital-`F` `Fix` instead of `fix`, since `fix` could be misconstrued for `Data.Function.fix`